File: sdks/novus-mundus-ts/src/calculators/progression.ts
/**
 * Progression Calculators
 *
 * XP, leveling, and progression calculations.
 */

import { getActivityMultiplier, TimeOfDay, ActivityType, getCurrentTimeOfDay } from './time';
import { applyBpsBonus } from './constants';

// ============================================================
// XP Requirements
// ============================================================

/**
 * XP required to reach a specific level (cumulative from level 1).
 *
 * Formula: 100 * 2.5^(level-2) for each level
 *
 * @param level - Target level
 * @returns Total XP required to reach that level
 *
 * @example
 * ```ts
 * xpRequiredForLevel(1);  // 0
 * xpRequiredForLevel(2);  // 100
 * xpRequiredForLevel(3);  // 350 (100 + 250)
 * xpRequiredForLevel(4);  // 975 (100 + 250 + 625)
 * ```
 */
export function xpRequiredForLevel(level: number): number {
  if (level <= 1) {
    return 0;
  }

  // Exponential formula: 100 * 2.5^(level-2)
  const base = 100;
  const multiplier = 2.5;
  const exponent = level - 2;

  return Math.floor(base * Math.pow(multiplier, exponent));
}

/**
 * XP required to go from one level to the next.
 *
 * @param currentLevel - Current level
 * @returns XP needed to reach next level
 */
export function xpToNextLevel(currentLevel: number): number {
  return xpRequiredForLevel(currentLevel + 1);
}

/**
 * Calculate cumulative XP required for a level (total from level 1).
 *
 * @param level - Target level
 * @returns Cumulative XP to reach that level
 */
export function cumulativeXpForLevel(level: number): number {
  let total = 0;
  for (let i = 2; i <= level; i++) {
    total += xpRequiredForLevel(i);
  }
  return total;
}

// ============================================================
// Level Calculations
// ============================================================

/**
 * Calculate the level for a given amount of XP.
 *
 * @param totalXp - Total XP accumulated
 * @returns Level reached
 */
export function levelFromXp(totalXp: number): number {
  let level = 1;
  let remainingXp = totalXp;

  while (level < 255) {
    const xpForNext = xpRequiredForLevel(level + 1);
    if (remainingXp < xpForNext) {
      break;
    }
    remainingXp -= xpForNext;
    level++;
  }

  return level;
}

/**
 * Calculate level and overflow XP.
 *
 * @param totalXp - Total XP accumulated
 * @returns [level, overflowXp] - Current level and XP toward next level
 */
export function levelAndOverflowFromXp(totalXp: number): [number, number] {
  let level = 1;
  let remainingXp = totalXp;

  while (level < 255) {
    const xpForNext = xpRequiredForLevel(level + 1);
    if (remainingXp < xpForNext) {
      break;
    }
    remainingXp -= xpForNext;
    level++;
  }

  return [level, remainingXp];
}

/**
 * Simulate granting XP and return new level and overflow.
 *
 * @param currentLevel - Current player level
 * @param currentXp - Current XP toward next level
 * @param xpAmount - XP to grant
 * @returns [newLevel, newXp, levelsGained]
 */
export function simulateGrantXp(
  currentLevel: number,
  currentXp: number,
  xpAmount: number
): [number, number, number] {
  let level = currentLevel;
  let xp = currentXp + xpAmount;
  let levelsGained = 0;

  while (level < 255) {
    const xpForNext = xpRequiredForLevel(level + 1);
    if (xp < xpForNext) {
      break;
    }
    xp -= xpForNext;
    level++;
    levelsGained++;
  }

  return [level, xp, levelsGained];
}

// ============================================================
// XP with Time Bonus
// ============================================================

/**
 * Calculate XP with time-of-day bonus.
 *
 * Golden hours (Dawn/Dusk) provide φ² (2.618x) XP bonus.
 * Night time provides φ (1.618x) bonus for wisdom.
 * Midday provides base 1.0x.
 *
 * @param baseXp - Base XP amount
 * @param timestamp - Current unix timestamp
 * @param longitude - Player longitude for time calculation
 * @param heroXpGainBps - Hero XP gain buff in basis points (optional)
 * @returns XP amount with time bonus applied
 */
export function calculateXpWithTimeBonus(
  baseXp: number,
  timestamp: number,
  longitude: number,
  heroXpGainBps: number = 0
): number {
  const timeOfDay = getCurrentTimeOfDay(timestamp, longitude);
  const xpMultiplier = getActivityMultiplier(ActivityType.XPGain, timeOfDay);
  let timeXp = Math.floor(baseXp * xpMultiplier);

  // Apply hero XP gain buff
  if (heroXpGainBps > 0) {
    timeXp = applyBpsBonus(timeXp, heroXpGainBps);
  }

  return timeXp;
}

/**
 * Calculate XP bonus multiplier for display purposes.
 *
 * @param timeOfDay - Current time of day
 * @returns Multiplier as a decimal (e.g., 2.618 for golden hours)
 */
export function getXpBonusMultiplier(timeOfDay: TimeOfDay): number {
  return getActivityMultiplier(ActivityType.XPGain, timeOfDay);
}

// ============================================================
// Progress Display Helpers
// ============================================================

/**
 * Calculate progress percentage to next level.
 *
 * @param currentLevel - Current player level
 * @param currentXp - Current XP toward next level
 * @returns Progress as a percentage (0-100)
 */
export function levelProgressPercent(currentLevel: number, currentXp: number): number {
  const xpForNext = xpRequiredForLevel(currentLevel + 1);
  if (xpForNext === 0) return 100;
  return Math.min(100, Math.floor((currentXp / xpForNext) * 100));
}

/**
 * Calculate XP remaining to next level.
 *
 * @param currentLevel - Current player level
 * @param currentXp - Current XP toward next level
 * @returns XP remaining to reach next level
 */
export function xpRemainingToNextLevel(currentLevel: number, currentXp: number): number {
  const xpForNext = xpRequiredForLevel(currentLevel + 1);
  return Math.max(0, xpForNext - currentXp);
}

/**
 * Format level progress as a string.
 *
 * @param currentLevel - Current player level
 * @param currentXp - Current XP toward next level
 * @returns Formatted string like "1,234 / 5,000 XP (24%)"
 */
export function formatLevelProgress(currentLevel: number, currentXp: number): string {
  const xpForNext = xpRequiredForLevel(currentLevel + 1);
  const percent = levelProgressPercent(currentLevel, currentXp);
  return `${currentXp.toLocaleString()} / ${xpForNext.toLocaleString()} XP (${percent}%)`;
}

// ============================================================
// XP Estimation
// ============================================================

/**
 * Estimate actions needed to level up.
 *
 * @param currentLevel - Current player level
 * @param currentXp - Current XP toward next level
 * @param xpPerAction - XP gained per action
 * @returns Number of actions needed to level up
 */
export function actionsToLevelUp(
  currentLevel: number,
  currentXp: number,
  xpPerAction: number
): number {
  if (xpPerAction <= 0) return Infinity;

  const remaining = xpRemainingToNextLevel(currentLevel, currentXp);
  return Math.ceil(remaining / xpPerAction);
}

/**
 * Estimate XP per hour at a given encounter rate.
 *
 * @param encountersPerHour - Number of encounters defeated per hour
 * @param averageRarity - Average encounter rarity (0-5)
 * @param timeOfDay - Current time of day for bonus
 * @returns Estimated XP per hour
 */
export function estimateXpPerHour(
  encountersPerHour: number,
  averageRarity: number,
  timeOfDay: TimeOfDay
): number {
  // Base XP by rarity
  const xpByRarity = [10, 25, 50, 100, 250, 500];
  const baseXp = xpByRarity[Math.min(Math.floor(averageRarity), 5)] ?? 10;

  // Apply time bonus
  const multiplier = getActivityMultiplier(ActivityType.XPGain, timeOfDay);
  const xpPerEncounter = Math.floor(baseXp * multiplier);

  return encountersPerHour * xpPerEncounter;
}

File: programs/novus_mundus/src/logic/progression.rs
use crate::{
    error::GameError,
    logic::{get_time_multiplier_bp, get_time_of_day, safe_math::apply_bp, ActivityType},
    state::PlayerAccount,
};

/// XP required to advance *into* `level` (the per-level cost, deducted on level-up).
///
/// Growth base is the golden ratio φ ≈ 1.618 — matching the rest of the
/// deterministic progression math, and keeping mid/late levels reachable for a
/// steady free player (a ×2.5 base put level 21 at ~6 billion cumulative XP).
/// Level 1->2: 100 XP
/// Level 2->3: 161 XP
/// Level 3->4: 261 XP
/// Level 4->5: 423 XP
pub fn xp_required_for_level(level: u8) -> u64 {
    // 100 * φ^(level-2), precomputed.
    XP_REQUIRED.get(level as usize).copied().unwrap_or(u64::MAX)
}

#[rustfmt::skip]
const XP_REQUIRED: [u64; 85] = [
    0, 0, 100, 161, 261, 423, 685, 1109,
    1794, 2903, 4697, 7601, 12299, 19900, 32199, 52100,
    84299, 136400, 220699, 357100, 577799, 934900, 1512699, 2447600,
    3960299, 6407900, 10368199, 16776100, 27144299, 43920400, 71064699, 114985100,
    186049799, 301034900, 487084699, 788119600, 1275204299, 2063323900, 3338528200, 5401852100,
    8740380300, 14142232400, 22882612700, 37024845100, 59907457800, 96932302900, 156839760700, 253772063600,
    410611824300, 664383887900, 1074995712200, 1739379600100, 2814375312300, 4553754912400, 7368130224700, 11921885137100,
    19290015361800, 31211900498900, 50501915860700, 81713816359600, 132215732220300, 213929548579900, 346145280800200, 560074829380101,
    906220110180302, 1466294939560403, 2372515049740705, 3838809989301108, 6211325039041814, 10050135028342922, 16261460067384738, 26311595095727664,
    42573055163112400, 68884650258840064, 111457705421952480, 180342355680792544, 291800061102745024, 472142416783537600, 763942477886282624, 1236084894669820416,
    2000027372556103168, 3236112267225923584, 5236139639782027264, 8472251907007950848, 13708391546789978112,
];

/// Grant XP to player with time-of-day bonus and handle level-ups
///
/// Golden hours (Dawn/Dusk) provide φ² (2.618x) XP bonus for enlightenment!
/// Night time provides √φ (1.272x) bonus for wisdom.
///
/// # Arguments
/// * `player` - Mutable reference to PlayerAccount
/// * `base_xp` - Base XP amount before time bonus
/// * `now` - Current unix timestamp
///
/// # Returns
/// (levels_gained, new_level, overflow_xp)
pub fn grant_xp_with_time_bonus(
    player: &mut PlayerAccount,
    base_xp: u64,
    now: i64,
) -> Result<(u8, u8, u64), GameError> {
    let time_of_day = get_time_of_day(now, player.current_long);
    let xp_bp = get_time_multiplier_bp(time_of_day, ActivityType::XPGain) as u64;
    let time_xp = apply_bp(base_xp, xp_bp).unwrap_or(base_xp);

    // Apply hero XP gain buff (multiplicative)
    // Formula: xp × (10000 + hero_xp_gain_bps) / 10000
    let xp_amount = if player.hero_xp_gain_bps() > 0 {
        let hero_multiplier = 10000u64.saturating_add(player.hero_xp_gain_bps() as u64);
        time_xp.saturating_mul(hero_multiplier) / 10000
    } else {
        time_xp
    };

    grant_xp(player, xp_amount)
}

/// Grant XP to player and handle level-ups (without time bonus)
/// Use grant_xp_with_time_bonus for time-aware XP grants.
/// Returns (levels_gained, new_level, overflow_xp)
pub fn grant_xp(player: &mut PlayerAccount, xp_amount: u64) -> Result<(u8, u8, u64), GameError> {
    let mut current_xp = player.current_xp.saturating_add(xp_amount);
    let mut current_level = player.level;
    let mut levels_gained = 0u8;

    // Check for level-ups. Guard the 255 cap BEFORE computing the next-level
    // cost: at current_level == 255, `current_level + 1` is a u8 overflow
    // (panics under debug overflow-checks, wraps to 0 in release).
    loop {
        if current_level >= 255 {
            break;
        }
        let xp_for_next = xp_required_for_level(current_level.saturating_add(1));

        if current_xp >= xp_for_next {
            current_xp = current_xp.saturating_sub(xp_for_next);
            current_level = current_level.saturating_add(1);
            levels_gained = levels_gained.saturating_add(1);
        } else {
            break;
        }
    }

    // Update player state
    player.current_xp = current_xp;
    player.level = current_level;

    // Update max stamina on level-up
    if levels_gained > 0 {
        crate::logic::update_max_stamina_for_tier(player);
    }

    Ok((levels_gained, current_level, current_xp))
}

/// Calculate XP rewards for various actions
pub fn calculate_xp_reward(action: XpAction) -> u64 {
    match action {
        XpAction::DefeatPlayer { target_level } => {
            // More XP for defeating higher-level players
            50u64.saturating_add((target_level as u64).saturating_mul(10))
        }
        XpAction::DefeatEncounter { rarity } => {
            // XP scales with encounter rarity
            match rarity {
                0 => 10,  // Common
                1 => 25,  // Uncommon
                2 => 50,  // Rare
                3 => 100, // Epic
                4 => 250, // Legendary
                5 => 500, // World Event
                _ => 0,
            }
        }
        XpAction::CompleteTravel { distance_km } => {
            // 1 XP per km traveled
            distance_km as u64
        }
        XpAction::CollectResources { amount } => {
            // 1 XP per 1000 resources collected
            amount / 1000
        }
    }
}

/// Actions that grant XP
pub enum XpAction {
    DefeatPlayer { target_level: u8 },
    DefeatEncounter { rarity: u8 },
    CompleteTravel { distance_km: u32 },
    CollectResources { amount: u64 },
}

/// Daily reward amounts (calculated with subscription tier multipliers)
pub struct DailyRewards {
    pub cash: u64,
    pub produce: u64,
    pub xp: u64,
}

/// Calculate daily rewards with subscription tier multipliers
///
/// Base values come from GameplayConfig (DAO-adjustable).
/// Multiplier comes from player's subscription tier (incentivizes higher tiers!).
///
/// # Arguments
/// * `player_tier` - Player's subscription tier (0-3)
/// * `gameplay_config` - GameplayConfig with base reward values
/// * `subscription_tiers` - Array of 4 subscription tier configs
///
/// # Returns
/// Daily reward amounts with tier multipliers applied
pub fn calculate_daily_rewards(
    player_tier: u8,
    gameplay_config: &crate::state::GameplayConfig,
    subscription_tiers: &[crate::state::SubscriptionTier; 4],
) -> Result<DailyRewards, GameError> {
    let base_cash = gameplay_config.daily_cash_base;
    let base_produce = gameplay_config.daily_produce_base;
    let base_xp = gameplay_config.daily_xp_base;

    // Get tier multiplier (tier 0-3, basis points: 10000 = 1.0x)
    let tier_index = player_tier.min(3) as usize;
    let tier = &subscription_tiers[tier_index];
    let multiplier = tier.daily_reward_multiplier;

    // Apply multiplier to all rewards (basis points: divide by 10000, no u128!)
    let cash = apply_bp(base_cash, multiplier as u64).ok_or(GameError::MathOverflow)?;

    let produce = apply_bp(base_produce, multiplier as u64).ok_or(GameError::MathOverflow)?;

    let xp = apply_bp(base_xp, multiplier as u64).ok_or(GameError::MathOverflow)?;

    Ok(DailyRewards { cash, produce, xp })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_xp_required() {
        // 100 * φ^(level-2)
        assert_eq!(xp_required_for_level(1), 0);
        assert_eq!(xp_required_for_level(2), 100);
        assert_eq!(xp_required_for_level(3), 161);
        assert_eq!(xp_required_for_level(4), 261);
    }

    #[test]
    fn xp_table_matches_formula() {
        // The fn must reproduce the old runtime formula bit-for-bit for every u8
        // level — anti-drift / paste-typo guard. Uses the same `libm::pow` the
        // table was generated from, so equality is exact. Testing the public fn
        // (not the raw table) also covers the u64::MAX fallback for level >= 85,
        // where the formula itself saturates `as u64`.
        for level in 0u16..=255 {
            let l = level as u8;
            let expected: u64 = if l <= 1 {
                0
            } else {
                (100.0 * libm::pow(crate::constants::PHI, (l as f64) - 2.0)) as u64
            };
            assert_eq!(xp_required_for_level(l), expected, "level {l}");
        }
    }
}
